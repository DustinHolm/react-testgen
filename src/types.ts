export interface ImportBlock {
    sourceFile: string,
    isInternal: boolean,
    imports: Import[]
}

export interface Import {
    isDefault: boolean,
    name: string,
    element?: ExportableElement
}

export interface Export {
    name: string,
    isDefault: boolean,
    element?: ExportableElement
}

export enum ExportType {
    Class,
    Function,
    Interface,
    Variable
}

export interface ExportableElement {
    type: ExportType,
    returnsJSX: boolean
}

export interface ClassElement extends ExportableElement {
    methods: MethodElement[],
    constructors: Constructor[]
}

export interface MethodElement extends FunctionElement {
    name: string
}

export interface FunctionElement extends ExportableElement {
    returnType: string,
    parameters: Attribute[]
}

export interface InterfaceElement extends ExportableElement {
    attributes: Attribute[]
}

export interface VariableElement extends ExportableElement {
    varType: string,
    varValue: string
}

export interface Attribute {
    name: string,
    type: string
}

export interface Constructor {
    parameters: Attribute[]
}