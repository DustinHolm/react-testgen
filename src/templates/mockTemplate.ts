import { Export, ExportType, FunctionElement, ImportBlock } from "../types"
import { createFunctionDefaults, createJSXDefaults } from "./defaultValuesTemplate"
import { supportedTypes, toPascalCase } from "./common"

export const createMock = (block: ImportBlock): string => {
    const jestActual = block.imports.find(i => 
        i.element === undefined
        || !supportedTypes.includes(i.element.type))
        ? `\
    ...jest.requireActual("${block.sourceFile}"),
`
        : ""

    const jestMocks = block.imports.map(i => createJestObject(i)).join("")

    const jestMockCall = `\
jest.mock("${block.sourceFile}", () => ({
${jestActual}${jestMocks}})
`

    const mocks = block.imports.map(i => createExportMock(i))

    return mocks.join("") + jestMockCall
}

const createJestObject = (importElement: Export): string => {
    if (importElement.isDefault) {
        return `\
    __esmodule: true,
    default: (...args: any[]) => mock${toPascalCase(importElement.name)}(args),
`
    } else {
        return `\
    ${importElement.name}: (...args: any[]) => mock${toPascalCase(importElement.name)}(args),
`
    }
}

const createExportMock = (exportElement: Export): string => {
    const mockName = "mock" + toPascalCase(exportElement.name)
    const element = exportElement.element

    switch (element?.type) {
        case ExportType.Function: return createFunctionMock(mockName, element as FunctionElement)
        // Other cases not yet implemented
        default: return ""
    }
}

const createFunctionMock = (name: string, functionElement: FunctionElement): string => {
    const defaults = functionElement.returnsJSX
        ? createJSXDefaults(name)
        : createFunctionDefaults(name, functionElement)

    return `\
let ${name}: jest.Mock
let ${name}Return: ${functionElement.returnType}
function reset${toPascalCase(name)}() {
${defaults}
}
`
}
