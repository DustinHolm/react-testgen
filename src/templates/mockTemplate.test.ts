import { Attribute, ClassElement, ExportType, FunctionElement, ImportBlock } from "../types"
import {createMock} from "./mockTemplate"

describe("mockTemplate Tests", () => {
    let importBlock: ImportBlock

    beforeEach(() => {
        resetMocks()
        importBlock = {
            sourceFile: "./mock",
            isInternal: true,
            imports: []
        }
    })

    test("With function based React Component", () => {
        importBlock.imports.push({
            name: "defaultTest",
            isDefault: true,
            element: getFunctionElement(true, 2)
        })

        const returnedString = createMock(importBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With class based React Component", () => {
        importBlock.imports.push({
            name: "defaultTest",
            isDefault: true,
            element: getClassElement(true, 2)
        })

        const returnedString = createMock(importBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With default function", () => {
        importBlock.imports.push({
            name: "defaultTest",
            isDefault: true,
            element: getFunctionElement(false, 1)
        })

        const returnedString = createMock(importBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With named function", () => {
        importBlock.imports.push({
            name: "namedTest",
            isDefault: false,
            element: getFunctionElement(false, 1)
        })

        const returnedString = createMock(importBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With multiple named functions", () => {
        importBlock.imports.push({
            name: "namedTest_1",
            isDefault: false,
            element: getFunctionElement(false, 1)
        })
        importBlock.imports.push({
            name: "namedTest_2",
            isDefault: false,
            element: getFunctionElement(false, 1)
        })

        const returnedString = createMock(importBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("With mixed functions", () => {
        importBlock.imports.push({
            name: "namedTest",
            isDefault: false,
            element: getFunctionElement(false, 1)
        })
        importBlock.imports.push({
            name: "defaultTest",
            isDefault: true,
            element: getFunctionElement(false, 1)
        })

        const returnedString = createMock(importBlock)

        expect(returnedString).toMatchSnapshot()
    })

    test("Calls createMockDefaults", () => {
        const exportElement = {
            name: "defaultTest",
            isDefault: true,
            element: getFunctionElement(false, 1)
        }
        importBlock.imports.push(exportElement)

        createMock(importBlock)

        expect(mockCreateFunctionDefaults).toHaveBeenCalledWith(["mockDefaultTest", exportElement.element])
    })
})

function getClassElement(returnsJSX: boolean, numberOfParameters: number): ClassElement {
    const parameters = []
    for (let i = 0; i < numberOfParameters; i++) {
        parameters.push(getAttribute())
    }
    
    return {
        type: ExportType.Class,
        returnsJSX: returnsJSX,
        methods: [{name: "testMethod", ...getFunctionElement(returnsJSX, 0)}],
        constructors: [{parameters: parameters}]
    }
}

function getFunctionElement(returnsJSX: boolean, numberOfParameters: number): FunctionElement {
    const parameters = []
    for (let i = 0; i < numberOfParameters; i++) {
        parameters.push(getAttribute())
    }
    
    return {
        type: ExportType.Function,
        returnsJSX: returnsJSX,
        returnType: "mockReturn",
        parameters: parameters
    }
}

function getAttribute(): Attribute {
    return {type: "mockType", name: "mockName"}
}

let mockCreateJSXDefaults: jest.Mock
function resetMockCreateJSXDefaults() {
    mockCreateJSXDefaults = jest.fn(() => "    mockedReset")
}

let mockCreateFunctionDefaults: jest.Mock
function resetMockCreateFunctionDefaults() {
    mockCreateFunctionDefaults = jest.fn(() => "    mockedReset")
}
jest.mock("./defaultValuesTemplate", () => ({
    createJSXDefaults: (...args: any[]) => mockCreateJSXDefaults(args),
    createFunctionDefaults: (...args: any[]) => mockCreateFunctionDefaults(args)
}))

function resetMocks() {
    resetMockCreateJSXDefaults()
    resetMockCreateFunctionDefaults()
}