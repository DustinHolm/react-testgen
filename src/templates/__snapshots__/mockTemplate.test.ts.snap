// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mockTemplate Tests With class based React Component 1`] = `
"let mockDefaultTest: jest.Mock
let mockDefaultTestReturn: JSX.Element
function resetMockDefaultTest() {
    mockedReset
}
jest.mock(\\"./mock\\", () => ({
    __esmodule: true,
    default: (...args: any[]) => mockDefaultTest(args),
}))
"
`;

exports[`mockTemplate Tests With default function 1`] = `
"let mockDefaultTest: jest.Mock
let mockDefaultTestReturn: mockReturn
function resetMockDefaultTest() {
    mockedReset
}
jest.mock(\\"./mock\\", () => ({
    __esmodule: true,
    default: (...args: any[]) => mockDefaultTest(args),
}))
"
`;

exports[`mockTemplate Tests With function based React Component 1`] = `
"let mockDefaultTest: jest.Mock
let mockDefaultTestReturn: JSX.Element
function resetMockDefaultTest() {
    mockedReset
}
jest.mock(\\"./mock\\", () => ({
    __esmodule: true,
    default: (...args: any[]) => mockDefaultTest(args),
}))
"
`;

exports[`mockTemplate Tests With mixed functions 1`] = `
"let mockNamedTest: jest.Mock
let mockNamedTestReturn: mockReturn
function resetMockNamedTest() {
    mockedReset
}
let mockDefaultTest: jest.Mock
let mockDefaultTestReturn: mockReturn
function resetMockDefaultTest() {
    mockedReset
}
jest.mock(\\"./mock\\", () => ({
    namedTest: (...args: any[]) => mockNamedTest(args),
    __esmodule: true,
    default: (...args: any[]) => mockDefaultTest(args),
}))
"
`;

exports[`mockTemplate Tests With multiple named functions 1`] = `
"let mockNamedTest_1: jest.Mock
let mockNamedTest_1Return: mockReturn
function resetMockNamedTest_1() {
    mockedReset
}
let mockNamedTest_2: jest.Mock
let mockNamedTest_2Return: mockReturn
function resetMockNamedTest_2() {
    mockedReset
}
jest.mock(\\"./mock\\", () => ({
    namedTest_1: (...args: any[]) => mockNamedTest_1(args),
    namedTest_2: (...args: any[]) => mockNamedTest_2(args),
}))
"
`;

exports[`mockTemplate Tests With named function 1`] = `
"let mockNamedTest: jest.Mock
let mockNamedTestReturn: mockReturn
function resetMockNamedTest() {
    mockedReset
}
jest.mock(\\"./mock\\", () => ({
    namedTest: (...args: any[]) => mockNamedTest(args),
}))
"
`;
